lkmpg.tex

```diff
@@ -1502,6 +1502,43 @@ \subsection{Flashing keyboard LEDs}
 Flashing keyboard LEDs can be such a solution: It is an immediate way to attract attention or to display a status condition.
 Keyboard LEDs are present on every hardware, they are always visible, they do not need any setup, and their use is rather simple and non-intrusive, compared to writing to a tty or a file.
 
+From v4.14 to v4.15, the timer API made a series of changes to improve memory safety.
+A buffer overflow in the area of a \cpp|timer_list| structure may be able to overwrite the \cpp|function| and \cpp|data| fields, providing attackers with a way to use return-object programming to call arbitrary functions within the kernel.
+Also, the prototype of the callback, with a single \cpp|unsigned long| argument, will let any type checking won't work and it is an obstacle to the \textit{control-flow integrity}.
+So, the timer callback should pass a pointer to the \cpp|timer_list| structure rather than an \cpp|unsigned long| argument.
+Then, wrap all the information the callback needs, including the \cpp|timer_list| structure, for simply using the \cpp|container_of| macro instead of the \cpp|unsigned long| value.
+
+Before Linux v4.14, \cpp|setup_timer| was used to initialize the timer and the \cpp|timer_list| structure looked like:
+\begin{code}
+struct timer_list {
+    unsigned long expires;
+    void (*function)(unsigned long);
+    unsigned long data;
+    u32 flags;
+    /* ... */
+};
+
+void setup_timer(struct timer_list *timer, void (*callback)(unsigned long),
+                 unsigned long data);
+\end{code}
+
+Since Linux v4.14, \cpp|timer_setup| is changed as following:
+\begin{code}
+void timer_setup(struct timer_list *timer,
+                 void (*callback)(struct timer_list *), unsigned int flags);
+\end{code}
+
+The \cpp|setup_timer| was removed at v4.15.
```

Comment:
Change "was removed" to "was then removed since Linux v4.15 for the sake of ..."